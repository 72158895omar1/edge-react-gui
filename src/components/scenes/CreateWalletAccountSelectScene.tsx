import { EdgeCurrencyWallet } from 'edge-core-js'
import * as React from 'react'
import { ActivityIndicator, ScrollView, View } from 'react-native'
import { cacheStyles } from 'react-native-patina'
import { sprintf } from 'sprintf-js'

import {
  createAccountTransaction,
  createCurrencyWallet,
  fetchAccountActivationInfo,
  fetchWalletAccountActivationPaymentInfo
} from '../../actions/CreateWalletActions'
import { CryptoIcon } from '../../components/icons/CryptoIcon'
import { WalletListModal, WalletListResult } from '../../components/modals/WalletListModal'
import { useHandler } from '../../hooks/useHandler'
import { lstrings } from '../../locales/strings'
import { getExchangeDenomination } from '../../selectors/DenominationSelectors'
import { config } from '../../theme/appConfig'
import { THEME } from '../../theme/variables/airbitz'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { EdgeSceneProps } from '../../types/routerTypes'
import { guessFromCurrencyCode } from '../../util/CurrencyInfoHelpers'
import { getWalletName } from '../../util/CurrencyWalletHelpers'
import { scale } from '../../util/scaling'
import { logEvent } from '../../util/tracking'
import { fixFiatCurrencyCode } from '../../util/utils'
import { SceneWrapper } from '../common/SceneWrapper'
import { PrimaryButton } from '../legacy/Buttons/PrimaryButton.ui'
import { FormattedText as Text } from '../legacy/FormattedText/FormattedText.ui'
import { Airship, showError } from '../services/AirshipInstance'
import { Theme, useTheme } from '../services/ThemeContext'

export interface AccountPaymentParams {
  requestedAccountName: string
  currencyCode: string
  ownerPublicKey: string
  activePublicKey: string
  requestedAccountCurrencyCode: string
}

interface Props extends EdgeSceneProps<'createWalletAccountSelect'> {}

export const CreateWalletAccountSelectScene = (props: Props) => {
  const { route } = props
  const { selectedFiat, selectedWalletType, accountName, existingWalletId } = route.params
  const dispatch = useDispatch()
  const theme = useTheme()
  const styles = getStyles(theme)

  const account = useSelector(state => state.core.account)
  const currencyWallets = useSelector(state => state.core.account.currencyWallets)
  const supportedAssets = useSelector(state => state.ui.createWallet.handleActivationInfo.supportedAssets)
  const activationCost = useSelector(state => state.ui.createWallet.handleActivationInfo.activationCost)
  const paymentCurrencyCode = useSelector(state => state.ui.createWallet.walletAccountActivationPaymentInfo.currencyCode)
  const amount = useSelector(state => state.ui.createWallet.walletAccountActivationPaymentInfo.amount)
  const existingCoreWallet = existingWalletId ? currencyWallets[existingWalletId] : undefined
  const paymentDenominationSymbol = useSelector(state =>
    paymentCurrencyCode != null && existingCoreWallet != null
      ? getExchangeDenomination(state, existingCoreWallet.currencyInfo.pluginId, paymentCurrencyCode).symbol ?? ''
      : ''
  )
  const walletAccountActivationQuoteError = useSelector(state => state.ui.createWallet.walletAccountActivationQuoteError)

  const [isCreatingWallet, setIsCreatingWallet] = React.useState(true)
  const [walletId, setWalletId] = React.useState('')

  const instructionSyntax = sprintf(
    lstrings.create_wallet_account_select_instructions_with_cost_4s,
    selectedWalletType.currencyCode,
    selectedWalletType.currencyCode,
    config.appNameShort,
    `${activationCost} ${selectedWalletType.currencyCode}`
  )
  const confirmMessageSyntax = sprintf(lstrings.create_wallet_account_make_payment, selectedWalletType.currencyCode)

  const { tokenId } = guessFromCurrencyCode(account, { currencyCode: selectedWalletType.currencyCode })

  const renameAndReturnWallet = useHandler(async (wallet: EdgeCurrencyWallet) => {
    await wallet.renameWallet(accountName)
    setIsCreatingWallet(false)
    return wallet
  })

  let createdWallet: Promise<EdgeCurrencyWallet>
  if (existingCoreWallet != null) {
    createdWallet = renameAndReturnWallet(existingCoreWallet)
  } else {
    createdWallet = dispatch(createCurrencyWallet(accountName, selectedWalletType.walletType, fixFiatCurrencyCode(selectedFiat.value))).then(wallet => {
      setIsCreatingWallet(false)
      return wallet
    })
  }

  const onPressSelect = useHandler(() => {
    Airship.show<WalletListResult>(bridge => (
      <WalletListModal bridge={bridge} navigation={props.navigation} headerTitle={lstrings.select_wallet} allowedAssets={supportedAssets} />
    ))
      .then(async ({ walletId, currencyCode }: WalletListResult) => {
        if (walletId && currencyCode) {
          await onSelectWallet(walletId, currencyCode)
        }
      })
      .catch(err => showError(err))
  })

  const onPressSubmit = useHandler(async () => {
    const createdWalletInstance = await createdWallet
    dispatch(createAccountTransaction(props.navigation, createdWalletInstance.id, accountName, walletId)).catch(err => showError(err))
  })

  const onSelectWallet = useHandler(async (walletId: string, paymentCurrencyCode: string) => {
    dispatch({ type: 'WALLET_ACCOUNT_ACTIVATION_ESTIMATE_ERROR', data: '' })
    setWalletId(walletId)
    const createdWalletInstance = await createdWallet
    const paymentInfo: AccountPaymentParams = {
      requestedAccountName: accountName,
      currencyCode: paymentCurrencyCode,
      ownerPublicKey: createdWalletInstance.publicWalletInfo.keys.ownerPublicKey,
      activePublicKey: createdWalletInstance.publicWalletInfo.keys.publicKey,
      requestedAccountCurrencyCode: selectedWalletType.currencyCode
    }
    dispatch(fetchWalletAccountActivationPaymentInfo(paymentInfo, createdWalletInstance))
  })

  React.useEffect(() => {
    logEvent('Activate_Wallet_Select')
    dispatch(fetchAccountActivationInfo(selectedWalletType.walletType)).catch(err => showError(err))
  }, [selectedWalletType.walletType, dispatch])

  const renderSelectWallet = () => {
    const currencyCode = selectedWalletType.currencyCode
    const isSelectWalletDisabled = !activationCost || activationCost === ''
    return (
      <View style={styles.selectPaymentLower}>
        <View style={styles.buttons}>
          <PrimaryButton disabled={isSelectWalletDisabled} style={styles.next} onPress={onPressSelect}>
            {isSelectWalletDisabled ? (
              <ActivityIndicator color={THEME.COLORS.ACCENT_MINT} />
            ) : (
              <PrimaryButton.Text>{lstrings.create_wallet_account_select_wallet}</PrimaryButton.Text>
            )}
          </PrimaryButton>
        </View>
        <View style={styles.paymentArea}>
          <Text style={styles.paymentLeft}>{lstrings.create_wallet_account_amount_due}</Text>
          <Text style={styles.paymentRight}>
            {activationCost} {currencyCode}
          </Text>
        </View>
      </View>
    )
  }

  const renderPaymentReview = () => {
    const wallet = account.currencyWallets[walletId]
    if (!wallet) return null
    const name = getWalletName(wallet)

    const isContinueButtonDisabled = isCreatingWallet

    return (
      <View>
        <View style={styles.selectPaymentLower}>
          <View style={styles.accountReviewWalletNameArea}>
            <Text style={styles.accountReviewWalletNameText}>
              {name}:{paymentCurrencyCode}
            </Text>
          </View>
          <View style={styles.paymentAndIconArea}>
            <View style={styles.paymentLeftIconWrap}>
              <CryptoIcon pluginId={wallet.currencyInfo.pluginId} sizeRem={1.5} />
            </View>
            <View style={styles.paymentArea}>
              <Text style={styles.paymentLeft}>
                {paymentDenominationSymbol} {amount} {paymentCurrencyCode}
              </Text>
              <Text style={styles.paymentRight}>
                {activationCost} {selectedWalletType.currencyCode}
              </Text>
            </View>
          </View>
        </View>
        <View style={styles.accountReviewInfoArea}>
          <Text style={styles.accountReviewInfoText}>
            {lstrings.create_wallet_crypto_type_label} {selectedWalletType.currencyCode}
          </Text>
          <Text style={styles.accountReviewInfoText}>
            {lstrings.create_wallet_fiat_type_label} {selectedFiat.label}
          </Text>
          <Text style={styles.accountReviewInfoText}>
            {lstrings.create_wallet_name_label} {accountName}
          </Text>
        </View>
        <View style={styles.accountReviewConfirmArea}>
          <Text style={styles.accountReviewConfirmText}>{lstrings.create_wallet_account_confirm}</Text>
        </View>
        <View style={styles.confirmButtonArea}>
          <PrimaryButton disabled={isContinueButtonDisabled} style={styles.confirmButton} onPress={onPressSubmit}>
            {/* we want it disabled with activity indicator if creating wallet, or wallet is created and pending quote */}
            {isContinueButtonDisabled ? (
              <ActivityIndicator color={THEME.COLORS.ACCENT_MINT} />
            ) : (
              <PrimaryButton.Text>{lstrings.legacy_address_modal_continue}</PrimaryButton.Text>
            )}
          </PrimaryButton>
        </View>
      </View>
    )
  }

  return (
    <SceneWrapper>
      <View style={styles.scene}>
        <ScrollView>
          <View style={styles.scrollableView}>
            <CryptoIcon marginRem={[1.5, 0, 0, 0]} sizeRem={4} tokenId={tokenId} />
            <View style={styles.createWalletPromptArea}>
              <Text style={styles.instructionalText}>{!walletId || walletAccountActivationQuoteError ? instructionSyntax : confirmMessageSyntax}</Text>
            </View>
            {!walletId || walletAccountActivationQuoteError ? renderSelectWallet() : renderPaymentReview()}
          </View>
          <View style={{ paddingBottom: 200 }} />
        </ScrollView>
      </View>
    </SceneWrapper>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  scene: {
    flex: 1,
    backgroundColor: THEME.COLORS.WHITE
  },
  scrollableView: {
    position: 'relative',
    paddingHorizontal: 20
  },
  createWalletPromptArea: {
    paddingTop: scale(16),
    paddingBottom: scale(8)
  },
  instructionalText: {
    fontSize: scale(16),
    textAlign: 'center',
    color: THEME.COLORS.GRAY_1
  },
  buttons: {
    marginTop: scale(24),
    flexDirection: 'row'
  },
  next: {
    marginLeft: scale(1),
    flex: 1
  },
  selectPaymentLower: {
    backgroundColor: THEME.COLORS.GRAY_4,
    width: '100%',
    marginVertical: scale(8),
    paddingHorizontal: scale(16)
  },
  paymentAndIconArea: {
    flexDirection: 'row'
  },
  paymentArea: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingVertical: scale(12),
    flex: 1
  },
  paymentLeft: {
    fontSize: scale(16),
    color: THEME.COLORS.GRAY_2
  },
  paymentLeftIconWrap: {
    paddingVertical: scale(12),
    marginRight: 6
  },

  paymentRight: {
    fontFamily: THEME.FONTS.BOLD,
    fontSize: scale(16),
    color: THEME.COLORS.GRAY_2
  },
  accountReviewWalletNameArea: {
    flexDirection: 'row',
    justifyContent: 'center',
    paddingTop: scale(14),
    paddingBottom: scale(8),
    alignItems: 'center'
  },
  accountReviewWalletNameText: {
    fontFamily: THEME.FONTS.BOLD,
    fontSize: scale(16),
    color: THEME.COLORS.SECONDARY
  },
  accountReviewInfoArea: {
    width: '100%',
    marginVertical: scale(10),
    paddingHorizontal: scale(10)
  },
  accountReviewInfoText: {
    color: THEME.COLORS.GRAY_2
  },
  accountReviewConfirmArea: {
    width: '100%',
    marginTop: scale(12),
    marginBottom: scale(12),
    paddingHorizontal: scale(10)
  },
  accountReviewConfirmText: {
    color: THEME.COLORS.GRAY_2,
    textAlign: 'center'
  },
  confirmButtonArea: {
    marginHorizontal: scale(30)
  },
  confirmButton: {}
}))
