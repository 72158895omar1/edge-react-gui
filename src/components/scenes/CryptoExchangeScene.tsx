import { gt, gte } from 'biggystring'
import { EdgeAccount, EdgeCurrencyWallet, EdgeSwapRequest, EdgeTokenId } from 'edge-core-js'
import * as React from 'react'
import { useState } from 'react'
import { Keyboard } from 'react-native'
import { sprintf } from 'sprintf-js'

import { selectWalletForExchange } from '../../actions/CryptoExchangeActions'
import { DisableAsset, ExchangeInfo } from '../../actions/ExchangeInfoActions'
import { updateMostRecentWalletsSelected } from '../../actions/WalletActions'
import { getSpecialCurrencyInfo } from '../../constants/WalletAndCurrencyConstants'
import { useSwapRequestOptions } from '../../hooks/swap/useSwapRequestOptions'
import { useHandler } from '../../hooks/useHandler'
import { lstrings } from '../../locales/strings'
import { useDispatch, useSelector } from '../../types/reactRedux'
import { EdgeSceneProps, NavigationBase } from '../../types/routerTypes'
import { emptyCurrencyInfo, GuiCurrencyInfo } from '../../types/types'
import { getTokenId, getWalletTokenId } from '../../util/CurrencyInfoHelpers'
import { getWalletName } from '../../util/CurrencyWalletHelpers'
import { zeroString } from '../../util/utils'
import { EdgeAnim, fadeInDown30, fadeInDown60, fadeInDown90, fadeInUp60, fadeInUp90 } from '../common/EdgeAnim'
import { SceneWrapper } from '../common/SceneWrapper'
import { WalletListModal, WalletListResult } from '../modals/WalletListModal'
import { Airship, showError, showWarning } from '../services/AirshipInstance'
import { cacheStyles, Theme, useTheme } from '../services/ThemeContext'
import { CryptoExchangeFlipInput } from '../themed/CryptoExchangeFlipInput'
import { ExchangedFlipInputAmounts } from '../themed/ExchangedFlipInput2'
import { LineTextDivider } from '../themed/LineTextDivider'
import { MiniButton } from '../themed/MiniButton'
import { SceneHeader } from '../themed/SceneHeader'
import { AlertCardUi4 } from '../ui4/AlertCardUi4'
import { ButtonsViewUi4 } from '../ui4/ButtonsViewUi4'

export interface SwapErrorDisplayInfo {
  message: string
  title: string
}

interface OwnProps extends EdgeSceneProps<'exchange'> {}

interface StateProps {
  account: EdgeAccount
  exchangeInfo: ExchangeInfo

  // The following props are used to populate the CryptoExchangeFlipInputs
  fromWalletInfo: FromWalletInfo
  toWalletInfo: ToWalletInfo
}

interface FromWalletInfo {
  fromWallet?: EdgeCurrencyWallet
  fromPluginId: string
  fromTokenId: EdgeTokenId
  fromWalletId: string
  fromWalletBalanceMap: Map<EdgeTokenId, string>
  fromWalletName: string
  fromWalletPrimaryInfo: GuiCurrencyInfo
  // Used to populate the confirmation modal
  fromCurrencyCode: string
}

interface ToWalletInfo {
  toWallet?: EdgeCurrencyWallet
  toTokenId: EdgeTokenId
  toWalletId: string
  toWalletName: string
  toWalletPrimaryInfo: GuiCurrencyInfo
  // Used to populate the confirmation modal
  toCurrencyCode: string
}

interface DispatchProps {
  onSelectWallet: (walletId: string, tokenId: EdgeTokenId, direction: 'from' | 'to') => Promise<void>
  getQuoteForTransaction: (navigation: NavigationBase, request: EdgeSwapRequest, onApprove: () => void) => void
}

type Props = OwnProps & StateProps & DispatchProps

interface State {
  whichWalletFocus: 'from' | 'to' // Which wallet FlipInput2 was last focused and edited
  fromAmountNative: string
  toAmountNative: string
  paddingBottom: number
}

const defaultFromWalletInfo: FromWalletInfo = {
  fromPluginId: '',
  fromTokenId: null,
  fromCurrencyCode: '',
  fromWalletBalanceMap: new Map<EdgeTokenId, string>(),
  fromWalletName: '',
  fromWalletPrimaryInfo: emptyCurrencyInfo,
  fromWalletId: ''
}

const defaultToWalletInfo: ToWalletInfo = {
  toTokenId: null,
  toCurrencyCode: '',
  toWalletName: '',
  toWalletPrimaryInfo: emptyCurrencyInfo,
  toWalletId: ''
}

const defaultState: State = {
  whichWalletFocus: 'from',
  fromAmountNative: '',
  toAmountNative: '',
  paddingBottom: 0
}

export const CryptoExchangeComponent = (props: Props) => {
  const { route } = props
  const { errorDisplayInfo } = route.params ?? {}
  const theme = useTheme()
  const styles = getStyles(theme)

  const [state, setState] = useState({
    ...defaultState
  })

  const checkDisableAsset = (disableAssets: DisableAsset[], walletId: string, guiCurrencyInfo: GuiCurrencyInfo): boolean => {
    const wallet = props.account.currencyWallets[walletId] ?? { currencyInfo: {} }
    const walletPluginId = wallet.currencyInfo.pluginId
    const walletTokenId = guiCurrencyInfo.tokenId ?? getTokenId(props.account, walletPluginId, guiCurrencyInfo.exchangeCurrencyCode)
    for (const disableAsset of disableAssets) {
      const { pluginId, tokenId } = disableAsset
      if (pluginId !== walletPluginId) continue
      if (tokenId === walletTokenId) return true
      if (tokenId === 'allCoins') return true
      if (tokenId === 'allTokens' && walletTokenId != null) return true
    }
    return false
  }

  const handleMax = () => {
    if (props.toWalletInfo.toWallet == null) {
      showWarning(`${lstrings.loan_select_receiving_wallet}`)
      Keyboard.dismiss()
      return
    }

    if (props.fromWalletInfo.fromWallet == null) {
      // Should never happen because max button UI is hidden unless a source wallet is selected
      throw new Error('No wallet selected')
    }

    const request: EdgeSwapRequest = {
      fromTokenId: props.fromWalletInfo.fromTokenId,
      fromWallet: props.fromWalletInfo.fromWallet,
      nativeAmount: '0',
      quoteFor: 'max',
      toTokenId: props.toWalletInfo.toTokenId,
      toWallet: props.toWalletInfo.toWallet
    }

    getQuote(request)
  }

  const handleNext = () => {
    if (props.fromWalletInfo.fromWallet == null || props.toWalletInfo.toWallet == null) {
      // Should never happen because next UI is hidden unless both source/destination wallets are selected
      throw new Error('No wallet selected')
    }

    const request: EdgeSwapRequest = {
      fromTokenId: props.fromWalletInfo.fromTokenId,
      fromWallet: props.fromWalletInfo.fromWallet,
      nativeAmount: state.whichWalletFocus === 'from' ? state.fromAmountNative : state.toAmountNative,
      quoteFor: state.whichWalletFocus,
      toTokenId: props.toWalletInfo.toTokenId,
      toWallet: props.toWalletInfo.toWallet
    }

    if (zeroString(request.nativeAmount)) {
      showError(`${lstrings.no_exchange_amount}. ${lstrings.select_exchange_amount}.`)
      return
    }

    if (checkExceedsAmount()) return

    getQuote(request)
  }

  const getQuote = (request: EdgeSwapRequest) => {
    const { exchangeInfo, navigation } = props
    if (exchangeInfo != null) {
      const disableSrc = checkDisableAsset(
        exchangeInfo.swap.disableAssets.source,
        props.fromWalletInfo.fromWalletId,
        props.fromWalletInfo.fromWalletPrimaryInfo
      )
      if (disableSrc) {
        showError(sprintf(lstrings.exchange_asset_unsupported, props.fromWalletInfo.fromWalletPrimaryInfo.exchangeCurrencyCode))
        return
      }

      const disableDest = checkDisableAsset(exchangeInfo.swap.disableAssets.destination, props.toWalletInfo.toWalletId, props.toWalletInfo.toWalletPrimaryInfo)
      if (disableDest) {
        showError(sprintf(lstrings.exchange_asset_unsupported, props.toWalletInfo.toWalletPrimaryInfo.exchangeCurrencyCode))
        return
      }
    }
    props.getQuoteForTransaction(navigation, request, resetState)
    Keyboard.dismiss()
  }

  const resetState = () => {
    setState(defaultState)
  }

  const checkExceedsAmount = (): boolean => {
    const { fromTokenId, fromWalletBalanceMap } = props.fromWalletInfo
    const { fromAmountNative, whichWalletFocus } = state
    const fromNativeBalance = fromWalletBalanceMap.get(fromTokenId) ?? '0'

    return whichWalletFocus === 'from' && gte(fromNativeBalance, '0') && gt(fromAmountNative, fromNativeBalance)
  }

  const launchFromWalletSelector = () => {
    renderDropUp('from')
  }

  const launchToWalletSelector = () => {
    renderDropUp('to')
  }

  const focusFromWallet = () => {
    setState({
      ...state,
      whichWalletFocus: 'from'
    })
  }

  const focusToWallet = () => {
    setState({
      ...state,
      whichWalletFocus: 'to'
    })
  }

  const fromAmountChanged = (amounts: ExchangedFlipInputAmounts) => {
    setState({
      ...state,
      fromAmountNative: amounts.nativeAmount
    })
  }

  const toAmountChanged = (amounts: ExchangedFlipInputAmounts) => {
    setState({
      ...state,
      toAmountNative: amounts.nativeAmount
    })
  }

  const renderButton = () => {
    const primaryNativeAmount = state.whichWalletFocus === 'from' ? state.fromAmountNative : state.toAmountNative
    const showNext = props.fromWalletInfo.fromCurrencyCode !== '' && props.toWalletInfo.toCurrencyCode !== '' && !!parseFloat(primaryNativeAmount)
    if (!showNext) return null
    if (checkExceedsAmount()) return null
    return <ButtonsViewUi4 primary={{ label: lstrings.string_next_capitalized, onPress: handleNext }} parentType="scene" />
  }

  const renderAlert = () => {
    const { fromPluginId, fromWalletBalanceMap, fromTokenId } = props.fromWalletInfo

    const { minimumPopupModals } = getSpecialCurrencyInfo(fromPluginId)
    const primaryNativeBalance = fromWalletBalanceMap.get(fromTokenId) ?? '0'

    if (minimumPopupModals != null && primaryNativeBalance < minimumPopupModals.minimumNativeBalance) {
      return <AlertCardUi4 title={lstrings.request_minimum_notification_title} body={minimumPopupModals.alertMessage} type="warning" />
    }

    if (errorDisplayInfo != null) {
      return <AlertCardUi4 title={errorDisplayInfo.title} body={errorDisplayInfo.message} type="error" />
    }

    if (checkExceedsAmount()) {
      return <AlertCardUi4 title={lstrings.exchange_insufficient_funds_title} body={lstrings.exchange_insufficient_funds_below_balance} type="error" />
    }

    return null
  }

  const renderDropUp = (whichWallet: 'from' | 'to') => {
    Airship.show<WalletListResult>(bridge => (
      <WalletListModal
        bridge={bridge}
        navigation={props.navigation}
        headerTitle={whichWallet === 'to' ? lstrings.select_recv_wallet : lstrings.select_src_wallet}
        showCreateWallet={whichWallet === 'to'}
        allowKeysOnlyMode={whichWallet === 'from'}
        filterActivation
      />
    ))
      .then(async result => {
        if (result?.type === 'wallet') {
          const { walletId, tokenId } = result
          await props.onSelectWallet(walletId, tokenId, whichWallet)
        }
      })
      .catch(error => showError(error))
  }

  const { fromWalletName } = props.fromWalletInfo
  const { toWalletName } = props.toWalletInfo

  const isFromFocused = state.whichWalletFocus === 'from'
  const isToFocused = state.whichWalletFocus === 'to'
  const fromHeaderText = sprintf(lstrings.exchange_from_wallet, fromWalletName)
  const toHeaderText = sprintf(lstrings.exchange_to_wallet, toWalletName)
  // Determines if a coin can have Exchange Max option
  const hasMaxSpend = getSpecialCurrencyInfo(props.fromWalletInfo.fromPluginId).noMaxSpend !== true

  return (
    <SceneWrapper hasTabs hasNotifications scroll keyboardShouldPersistTaps="handled" padding={theme.rem(0.5)}>
      <EdgeAnim style={styles.header} enter={fadeInUp90}>
        <SceneHeader title={lstrings.title_exchange} underline />
      </EdgeAnim>
      <EdgeAnim enter={fadeInUp60}>
        <CryptoExchangeFlipInput
          wallet={props.fromWalletInfo.fromWallet}
          buttonText={lstrings.select_src_wallet}
          headerText={fromHeaderText}
          primaryCurrencyInfo={props.fromWalletInfo.fromWalletPrimaryInfo}
          overridePrimaryNativeAmount={state.fromAmountNative}
          launchWalletSelector={launchFromWalletSelector}
          onCryptoExchangeAmountChanged={fromAmountChanged}
          isFocused={isFromFocused}
          focusMe={focusFromWallet}
          onNext={handleNext}
        >
          {hasMaxSpend ? <MiniButton label={lstrings.string_max_cap} marginRem={[0.5, 0, 0.75]} onPress={handleMax} alignSelf="center" /> : null}
        </CryptoExchangeFlipInput>
      </EdgeAnim>
      <EdgeAnim>
        <LineTextDivider title={lstrings.string_to_capitalize} lowerCased />
      </EdgeAnim>
      <EdgeAnim enter={fadeInDown30}>
        <CryptoExchangeFlipInput
          wallet={props.toWalletInfo.toWallet}
          buttonText={lstrings.select_recv_wallet}
          headerText={toHeaderText}
          primaryCurrencyInfo={props.toWalletInfo.toWalletPrimaryInfo}
          overridePrimaryNativeAmount={state.toAmountNative}
          launchWalletSelector={launchToWalletSelector}
          onCryptoExchangeAmountChanged={toAmountChanged}
          isFocused={isToFocused}
          focusMe={focusToWallet}
          onNext={handleNext}
        />
      </EdgeAnim>
      <EdgeAnim enter={fadeInDown60}>{renderAlert()}</EdgeAnim>
      <EdgeAnim enter={fadeInDown90}>{renderButton()}</EdgeAnim>
    </SceneWrapper>
  )
}

const getStyles = cacheStyles((theme: Theme) => ({
  mainScrollView: {
    flex: 1
  },
  header: {
    marginLeft: -theme.rem(0.5),
    width: '100%',
    marginVertical: theme.rem(1)
  },
  scrollViewContentContainer: {
    alignItems: 'center',
    marginHorizontal: theme.rem(0.5)
  }
}))

export const CryptoExchangeScene = (props: OwnProps) => {
  const dispatch = useDispatch()
  const { navigation, route } = props

  const account = useSelector(state => state.core.account)
  const currencyWallets = useSelector(state => state.core.account.currencyWallets)
  const cryptoExchange = useSelector(state => state.cryptoExchange)
  const exchangeInfo = useSelector(state => state.ui.exchangeInfo)

  const swapRequestOptions = useSwapRequestOptions()

  const { fromWalletId, toWalletId } = cryptoExchange

  let fromWalletInfo = defaultFromWalletInfo
  let toWalletInfo = defaultToWalletInfo

  if (fromWalletId != null && currencyWallets[fromWalletId] != null) {
    const fromWallet = currencyWallets[fromWalletId]
    const { fromWalletPrimaryInfo } = cryptoExchange
    const { exchangeCurrencyCode } = fromWalletPrimaryInfo
    const fromTokenId = getWalletTokenId(fromWallet, exchangeCurrencyCode)

    const fromWalletName = getWalletName(fromWallet)
    const { balanceMap: fromWalletBalanceMap } = fromWallet

    fromWalletInfo = {
      fromWallet,
      fromTokenId,
      fromWalletId,
      fromWalletName,
      fromWalletBalanceMap,
      fromCurrencyCode: exchangeCurrencyCode,
      fromWalletPrimaryInfo,
      fromPluginId: fromWallet.currencyInfo.pluginId
    }
  }

  // Get the values of the 'To' Wallet
  if (toWalletId != null && currencyWallets[toWalletId] != null) {
    const toWallet = currencyWallets[toWalletId]
    const { toWalletPrimaryInfo } = cryptoExchange
    const { exchangeCurrencyCode } = toWalletPrimaryInfo
    const toTokenId = getWalletTokenId(toWallet, exchangeCurrencyCode)

    const toWalletName = getWalletName(toWallet)

    toWalletInfo = {
      toWallet,
      toTokenId,
      toWalletId,
      toWalletName,
      toCurrencyCode: exchangeCurrencyCode,
      toWalletPrimaryInfo
    }
  }

  const handleSelectWallet = useHandler(async (walletId: string, tokenId: EdgeTokenId, direction: 'from' | 'to') => {
    await dispatch(selectWalletForExchange(walletId, tokenId, direction))
    dispatch(updateMostRecentWalletsSelected(walletId, tokenId))
  })

  const handleGetQuoteForTransaction = useHandler((navigation: NavigationBase, swapRequest: EdgeSwapRequest, onApprove: () => void) => {
    navigation.navigate('exchangeQuoteProcessing', {
      swapRequest,
      swapRequestOptions,
      onCancel: () => {
        navigation.goBack()
      },
      onDone: quotes => {
        navigation.replace('exchangeQuote', {
          selectedQuote: quotes[0],
          quotes,
          onApprove
        })
      }
    })
  })

  return (
    <CryptoExchangeComponent
      route={route}
      onSelectWallet={handleSelectWallet}
      getQuoteForTransaction={handleGetQuoteForTransaction}
      navigation={navigation}
      account={account}
      exchangeInfo={exchangeInfo}
      fromWalletInfo={fromWalletInfo}
      toWalletInfo={toWalletInfo}
    />
  )
}
