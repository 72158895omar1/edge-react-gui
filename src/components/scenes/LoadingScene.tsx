import * as React from 'react'
import { ActivityIndicator, StyleSheet, View } from 'react-native'

import { SceneWrapper } from '../common/SceneWrapper'
import { useTheme } from '../services/ThemeContext'

export const LoadingScene = () => {
  const theme = useTheme()
  return (
    <SceneWrapper background="theme" hasHeader={false} hasTabs={false}>
      <View style={styles.container}>
        <ActivityIndicator color={theme.loadingIcon} size="large" />
      </View>
    </SceneWrapper>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  }
})
