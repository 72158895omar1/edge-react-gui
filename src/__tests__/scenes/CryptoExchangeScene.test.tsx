import { describe, expect, it } from '@jest/globals'
import * as React from 'react'
import TestRenderer from 'react-test-renderer'

import { CryptoExchangeComponent } from '../../components/scenes/CryptoExchangeScene'
import { initialState } from '../../reducers/ExchangeInfoReducer'
import { FakeProviders, FakeState } from '../../util/fake/FakeProviders'
import { fakeRootState } from '../../util/fake/fakeRootState'
import { fakeSceneProps } from '../../util/fake/fakeSceneProps'

describe('CryptoExchangeComponent', () => {
  it('should render with loading props', () => {
    const rootState: FakeState = { ...fakeRootState }

    const renderer = TestRenderer.create(
      <FakeProviders initialState={rootState}>
        <CryptoExchangeComponent
          {...fakeSceneProps('exchange', {})}
          exchangeInfo={initialState}
          onSelectWallet={async () => undefined}
          getQuoteForTransaction={() => undefined}
        />
      </FakeProviders>
    )

    expect(renderer.toJSON()).toMatchSnapshot()
    renderer.unmount()
  })
})
