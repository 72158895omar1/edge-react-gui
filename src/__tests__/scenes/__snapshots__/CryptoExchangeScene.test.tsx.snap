// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CryptoExchangeComponent should render with loading props 1`] = `
<View
  style={
    {
      "flex": 1,
      "marginHorizontal": 11,
    }
  }
>
  <View
    style={
      {
        "marginLeft": -11,
        "marginVertical": 22,
        "width": "100%",
      }
    }
  >
    <Memo(SceneHeaderComponent)
      title="Exchange"
      underline={true}
    />
  </View>
  <Memo(Connect(WithTheme(CryptoExchangeFlipInputWrapperComponent)))
    buttonText="Select Source Wallet"
    focusMe={[Function]}
    headerText="Exchange from BTC Wallet"
    isFocused={true}
    launchWalletSelector={[Function]}
    onCryptoExchangeAmountChanged={[Function]}
    onNext={[Function]}
    overridePrimaryNativeAmount=""
    primaryCurrencyInfo={
      {
        "displayCurrencyCode": "BTC",
        "exchangeCurrencyCode": "BTC",
      }
    }
    walletId=""
  >
    <MiniButton
      alignSelf="center"
      label="MAX"
      marginRem={
        [
          0.5,
          0,
          1,
        ]
      }
      onPress={[Function]}
    />
  </Memo(Connect(WithTheme(CryptoExchangeFlipInputWrapperComponent)))>
  <Memo(LineTextDividerComponent)
    lowerCased={true}
    title="To"
  />
  <Memo(Connect(WithTheme(CryptoExchangeFlipInputWrapperComponent)))
    buttonText="Select Receiving Wallet"
    focusMe={[Function]}
    headerText="Exchange to ETH Wallet"
    isFocused={false}
    launchWalletSelector={[Function]}
    onCryptoExchangeAmountChanged={[Function]}
    onNext={[Function]}
    overridePrimaryNativeAmount=""
    primaryCurrencyInfo={
      {
        "displayCurrencyCode": "ETH",
        "displayDenomination": {
          "multiplier": "100000000",
          "name": "Ethereum",
        },
        "exchangeCurrencyCode": "ETH",
        "exchangeDenomination": {
          "multiplier": "10000000",
          "name": "Gwei",
        },
      }
    }
    walletId=""
  />
</View>
`;
